src/os/bluestore/BlueStore.cc

```diff
@@ -14516,12 +14560,14 @@ void BlueStore::_do_truncate(
   if (offset == o->onode.size)
     return;
 
+  int64_t ondisk_offset = 0;
   if (offset < o->onode.size) {
     WriteContext wctx;
     uint64_t length = o->onode.size - offset;
     o->extent_map.fault_range(db, offset, length);
     o->extent_map.punch_hole(c, offset, length, &wctx.old_extents);
     o->extent_map.dirty_range(offset, length);
+    ondisk_offset = wctx.old_extents.begin()->r.begin()->offset;
```

Comment:
```
2020-07-27T09:21:27.565 INFO:journalctl@ceph.osd.0.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50694]: terminate called after throwing an instance of 'boost::wrapexcept<boost::bad_get>'
2020-07-27T09:21:27.565 INFO:journalctl@ceph.osd.0.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50694]:   what():  boost::bad_get: failed value get using boost::get
2020-07-27T09:21:27.565 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]: *** Caught signal (Segmentation fault) **
2020-07-27T09:21:27.565 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  in thread 7facf5fce700 thread_name:tp_osd_tp
2020-07-27T09:21:27.566 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  ceph version 16.0.0-3845-g6ed28664f92 (6ed28664f92a0636259d1a756597732c441fbce4) pacific (dev)
2020-07-27T09:21:27.566 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  1: (()+0x12dd0) [0x7fad1febbdd0]
2020-07-27T09:21:27.567 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  2: (BlueStore::_do_truncate(BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection
>&, boost::intrusive_ptr<BlueStore::Onode>, unsigned long, std::set<BlueStore::SharedBlob*, std::less<BlueStore::SharedBlob*>, std::allocator<BlueStore::SharedBlob*> >*)+0x389) [0x561db5fdaa49]
2020-07-27T09:21:27.567 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  3: (BlueStore::_truncate(BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection>&,
 boost::intrusive_ptr<BlueStore::Onode>&, unsigned long)+0x80) [0x561db5fdae60]
2020-07-27T09:21:27.567 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  4: (BlueStore::_txc_add_transaction(BlueStore::TransContext*, ceph::os::Transaction*)+0x1a4a) [0
x561db600075a]
2020-07-27T09:21:27.568 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  5: (BlueStore::queue_transactions(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, std::vecto
r<ceph::os::Transaction, std::allocator<ceph::os::Transaction> >&, boost::intrusive_ptr<TrackedOp>, ThreadPool::TPHandle*)+0x317) [0x561db6001b37]
2020-07-27T09:21:27.568 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  6: (non-virtual thunk to PrimaryLogPG::queue_transactions(std::vector<ceph::os::Transaction, std
::allocator<ceph::os::Transaction> >&, boost::intrusive_ptr<OpRequest>)+0x58) [0x561db5c95238]
2020-07-27T09:21:27.568 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  7: (ReplicatedBackend::submit_transaction(hobject_t const&, object_stat_sum_t const&, eversion_t const&, std::unique_ptr<PGTransaction, std::default_delete<PGTransaction> >&&, eversion_t const&, eversion_t const&, std::vector<pg_log_entry_t, std::allocator<pg_log_entry_t> >&&, std::optional<pg_hit_set_history_t>&, Context*, unsigned long, osd_reqid_t, boost::intrusive_ptr<OpRequest>)+0xc9d) [0x561db5e47c9d]
2020-07-27T09:21:27.569 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  8: (PrimaryLogPG::issue_repop(PrimaryLogPG::RepGather*, PrimaryLogPG::OpContext*)+0xce0) [0x561db5c01420]
2020-07-27T09:21:27.569 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  9: (PrimaryLogPG::execute_ctx(PrimaryLogPG::OpContext*)+0x115d) [0x561db5c5c35d]
2020-07-27T09:21:27.570 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  10: (PrimaryLogPG::do_op(boost::intrusive_ptr<OpRequest>&)+0x2d46) [0x561db5c64e06]
2020-07-27T09:21:27.570 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  11: (PrimaryLogPG::do_request(boost::intrusive_ptr<OpRequest>&, ThreadPool::TPHandle&)+0xd40) [0x561db5c6c170]
2020-07-27T09:21:27.570 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  12: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x309) [0x561db5aebfc9]
2020-07-27T09:21:27.571 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  13: (ceph::osd::scheduler::PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x68) [0x561db5d51ca8]
2020-07-27T09:21:27.571 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  14: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x12ef) [0x561db5b0b1bf]
2020-07-27T09:21:27.572 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  15: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x5c4) [0x561db6150954]
2020-07-27T09:21:27.572 INFO:journalctl@ceph.osd.1.smithi041.stdout:Jul 27 09:21:27 smithi041 bash[50727]:  16: (ShardedThreadPool::WorkThreadSharded::entry()+0x14) [0x561db61535b4]
```

see https://pulpito.ceph.com/kchai-2020-07-27_09:02:33-rados-wip-kefu-testing-2020-07-27-1351-distro-basic-smithi/5261438/

i am going to drop this PR and rerun the test. if the test still fails, i will withdraw my nack.

Reply:
@tchaikov Can you merge this please?  Thanks!

Reply:
@agayev sure, but i need to test your change first to verify it does not create any regressions.

Reply:
@tchaikov Of course :-).  By "merge" I meant everything that you usually do for merging.